docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+
+    Another option is a preprocessing step before type inference that rewrites the Flux AST by replacing calls to `keep` and `drop` with their `map` equivalents.
+
+* aggregates
+
+    Assigning accurate types to aggregates is also a bit challenging.
+    This is due to the fact the aggregates will implicitly drop columns that are not part of the group key.
+    However if aggregates didn't drop columns dynamically, we could type them like so:
+    ```
+    forall ['r] (column: str, tables: [ {_time: time | 'r} ]) -> [ 'r ]
+    ```
+    In words, aggregates take a list of records with a `_time` fields and return that same list minus the `_time` field.
+
+Note that functions which remove columns from tables dynamically pose challenges to type inference.
+A generic type can still be given to the function but any information we have on the input, we lose on the output.
+`join` has the same problem, but for a different reason.
+Conceptually `join` just concatenates records.
+Unfortunately record concatenation is not supported by the type system and as a result all we can say about join is that it produces a generic list of records.
+Hence any information we had about the records going into join, we lose going out.
+
+#### What about group?
+
+Group presents a different challenge.
+While we can think of streams logically as a list of heterogeneous records, internally they are represented using a much more efficient columnar implementation.
+So while the type signature of `group` is just:
+```
+forall ['r] where 'r:Rec (columns: [str], mode: str, tables: ['r]) -> ['r]
+```
+grouping can potentially place data of two different types in the same column, which will result in a type error at runtime.
+
+A potential way to avoid this would be to create a new _kind_ describing any record whose type is known for every possible key.
+In other words, a record _kind_ that enforces schema homogeneity.
+As a result a query like `from() |> range() |> group(columns: ["host"])` would fail unification as the type of the `_value` column produced by `from` is unknown.
+This might be too restrictive though.
+
+#### What about pivot?
+
+The same strategy we employed for `group` we can also employ for `pivot`.
+This would give us partial information on the type of pivot's output.
+However this too might be too restrictive.
+Do users want to be able to group and pivot heterogeneous data?
+
+#### What about other data sources?
+
+Unlike InfluxDB, sources like `csv` and `sql` cannot have such a strict type signature.
+Instead sources like these need to be defined to return lists of generic records `forall ['r] (...) -> [ 'r ]`.
+However based on usage, type inference will be able to determine a more specific type for `'r`.
+Each source can then perform their own type checking to ensure the data on disk matches with the type of `'r` that is inferred.
```

Comment:
Two thoughts here:

### Filter vs Type Check

 If we use type inference as a filter back to sources it solves a lot of problems and creates a few of its own.

 First, you are guaranteed to get semi-homogenous data out of the source, by semi-homogenous I mean that any field that we know about will have the expected type, but other fields may still exist. This solves many of the issues we were having with heterogeneous data. 

Second, it creates a problem where users may inadvertently filter their data instead of getting an error. In your previous example of using: `r.host == 1` in a filter predicate you would get an error because `from` knows that the other columns are of type string so a type error is produced and the user can fix their error. But as you point out for other sources this would result is data that does not have an integer `host` field to be filtered, resulting in no data being returned and the user being confused about why no data is read from the `csv` file. 


### Query by Example

The idea of query by example is the user provides a query that demonstrates the entire structure of the data they expect. This is what @aanthony1243 was discussing at the last all-hands. The idea is that the syntax and arguments of the `from` function (or whatever wrapper we put on it) provides enough information to get homogeneous records from sources. 

This seems to have the advantage of disambiguating a users intention. Meaning in the above case if the user has a typo related to types in such a query it would not be as surprising to the user that the data was filtered since the expectation was the query types were filters. 

In short, maybe these two approaches are the same thing and we just need to correctly educate users that types are filters?

Reply:
Having `from` implicitly filter rows based on the inferred types is heavily influenced by @aanthony1243's query by example proposal. And in the case of the type signature I propose for `from`, this would only apply to the `_value` columns as that is the only column for which types are unknown.

Note however that this really only applies to the influxDB source. Others like csv and sql I don't think should do any implicit filtering. Instead the sources themselves should implement type checking and if a type error occurs it happens at the source instead of halfway through the pipeline.

Also keep in mind that having `from` implicitly filter on types and even updating the current transformations with accurate type signatures does **not** fall under this proposal. It is something that we can do easily and iteratively after simplifying the _kind_ system and introducing extensible record types. But only the latter two are required as part of this 2nd iteration.